torchmetrics/image/map.py

```diff
@@ -0,0 +1,369 @@
+        if any(k not in p for p in preds):
+            raise ValueError(f"Expected all dicts in `preds` to contain the `{k}` key")
+
+    for k in ["boxes", "labels"]:
+        if any(k not in p for p in targets):
+            raise ValueError(f"Expected all dicts in `target` to contain the `{k}` key")
+
+    if any(type(pred["boxes"]) is not torch.Tensor for pred in preds):
+        raise ValueError("Expected all boxes in `preds` to be of type torch.Tensor")
+    if any(type(pred["scores"]) is not torch.Tensor for pred in preds):
+        raise ValueError("Expected all scores in `preds` to be of type torch.Tensor")
+    if any(type(pred["labels"]) is not torch.Tensor for pred in preds):
+        raise ValueError("Expected all labels in `preds` to be of type torch.Tensor")
+    if any(type(target["boxes"]) is not torch.Tensor for target in targets):
+        raise ValueError("Expected all boxes in `target` to be of type torch.Tensor")
+    if any(type(target["labels"]) is not torch.Tensor for target in targets):
+        raise ValueError("Expected all labels in `target` to be of type torch.Tensor")
+
+
+class MAP(Metric):
+    r"""
+    Computes the `Mean-Average-Precision (mAP) and Mean-Average-Recall (mAR)`_ for object detection predictions.
+    Optionally, the mAP and mAR values can be calculated per class.
+
+    Predicted boxes and targets have to be in Pascal VOC format
+    (xmin-top left, ymin-top left, xmax-bottom right, ymax-bottom right).
+    See the `update` function for more information about the input format to this metric.
+
+    .. note::
+        This metric is a wrapper for the
+        `pycocotools <https://github.com/cocodataset/cocoapi/tree/master/PythonAPI/pycocotools>`_,
+        which is a standard implementation for the mAP metric for object detection. Using this metric
+        therefore requires you to have `pycocotools` installed. Please install with `pip install pycocotools` or
+        `pip install torchmetrics[image]`
+
+    .. note::
+        As the pycocotools library cannot deal with tensors directly, all results have to be transfered
+        to the CPU, this might have an performance impact on your training
+
+    Args:
+        num_classes:
+            Number of classes, required for mAP values per class. default: 0 (deactivate)
+        compute_on_step:
+            Forward only calls ``update()`` and return None if this is set to False. default: False
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step. default: False
+        process_group:
+            Specify the process group on which synchronization is called. default: None (which selects the entire world)
+
+    Raises:
+        ValueError:
+            If ``pycocotools`` is not installed
+        ValueError:
+            If ``num_classes`` is not an integer larger or equal to 0
+    """
+
+    def __init__(
```

Comment:
Can we simplify this:

    def __init__(self, num_classes: int = 0, **kwargs) -> None:
        super().__init__(**kwargs)

Reply:
LGTM.